### ISSUE TITLE
KDE in polar plot

### ISSUE THREAD

**Issue by ClementCaporal** [LaboratoryOpticsBiosciences/napari-brainbow-diagnose#33]
This issue will detail several ideas I tried to create a polar plot with a KDE.
I couldn't find a satisfactory solution, but I will list here the following attemps:
- scipy `gaussian_kde`
- sklearn `KernelDensity` with metrics `haversine` and `pyfunc` (aka custom metric)

**Comment by ClementCaporal** [comment]
## SCIPY gaussian_kde

```python
import matplotlib.pyplot as plt
from scipy.stats import gaussian_kde
import numpy as np

# Example data points in Cartesian coordinates
theta = np.random.rand(100) * 2 * np.pi
r = np.random.rand(100)

# Create a meshgrid for polar coordinates
radii = np.linspace(0, 1, 100)
angles = np.linspace(0, 2 * np.pi, 360, endpoint=False)
R, Theta = np.meshgrid(radii, angles)

kde = gaussian_kde(np.vstack([theta, r]))
density = kde(np.vstack([Theta.ravel(), R.ravel()]))
density = density.reshape(R.shape)

fig, ax = plt.subplots(subplot_kw={'projection': 'polar'})
ax.pcolormesh(Theta, R, density)
plt.show()
```

![Image](https://github.com/user-attachments/assets/a9ae8b65-786e-4aca-99db-50f3d771d1b3)




This will create a border artefact around the 0° and 359°

So we need a method that will wrap the angles

**Comment by ClementCaporal** [comment]
## sklearn with haversine

```python
import matplotlib.pyplot as plt
from sklearn.neighbors import KernelDensity

import numpy as np

# Example data points in Cartesian coordinates
theta = np.random.rand(100) * 2 * np.pi
r = np.random.rand(100)

# Create a meshgrid for polar coordinates
radii = np.linspace(0, 1, 100)
angles = np.linspace(0, 2 * np.pi, 360, endpoint=False)
R, Theta = np.meshgrid(radii, angles)
grid_points = np.vstack([Theta.ravel(), R.ravel()])

kde = KernelDensity(kernel='gaussian', metric='haversine', bandwidth=0.1)
kde.fit(np.vstack([theta, r]).T)
log_density = kde.score_samples(grid_points.T)
log_density = log_density.reshape(R.shape)
density = np.exp(log_density)

fig, ax = plt.subplots(subplot_kw={'projection': 'polar'})
ax.pcolormesh(Theta, R, log_density)
ax.plot(theta, r, 'r.', alpha=0.1)
plt.show()
```

haversine will warp both the hue and saturation creating these rays

**Comment by ClementCaporal** [comment]
## sklearn with custom metric

```python
import matplotlib.pyplot as plt
from sklearn.neighbors import KernelDensity

import numpy as np

# Example data points in Cartesian coordinates
theta = np.random.rand(100) * 2 * np.pi
r = np.random.rand(100)

# Create a meshgrid for polar coordinates
radii = np.linspace(0, 1, 100)
angles = np.linspace(0, 2 * np.pi, 360, endpoint=False)
R, Theta = np.meshgrid(radii, angles)
grid_points = np.vstack([Theta.ravel(), R.ravel()])

def hue_saturation_distance(x, y, w_h=1, w_s=1):
    h1, s1 = x
    h2, s2 = y

    # Circular distance for hue
    hue_diff = abs(h1 - h2) % (2 * np.pi)
    hue_diff = min(hue_diff, 2 * np.pi - hue_diff)

    # Linear distance for saturation
    saturation_diff = (s1 - s2) ** 2

    # Combined weighted distance
    return np.sqrt(w_h * hue_diff + w_s * saturation_diff)

kde = KernelDensity(kernel='gaussian', metric='pyfunc', metric_params={"func":hue_saturation_distance}, bandwidth=0.1)
kde.fit(np.vstack([theta, r]).T)
log_density = kde.score_samples(grid_points.T)
log_density = log_density.reshape(R.shape)
density = np.exp(log_density)

fig, ax = plt.subplots(subplot_kw={'projection': 'polar'})
ax.pcolormesh(Theta, R, density)
ax.plot(theta, r, 'r.', alpha=1)
plt.show()
```


![Image](https://github.com/user-attachments/assets/6ccc0176-8066-4e20-af63-54fcbe260160)


`bandwidth` could be tuned but want is stricking is the local radial dispersion that I don't understand

**Comment by ClementCaporal** [comment]
Oh!
It is because `hue_diff` should be squared....

![Image](https://github.com/user-attachments/assets/908ca046-3a56-4d73-8c79-4ba3642b3a63)

It is now solved!
